### PR TITLE
Add semantic versioning

### PR DESCRIPTION
**Pull request by js6pak**
## Description
Adds semantic versioning to both BepInEx and plugins.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: BepInEx.Bootstrap/BepInEx.Bootstrap.csproj [hunk 1]
```diff
@@ -7,10 +7,6 @@
     <OutputPath>..\bin\patcher\</OutputPath>
     <LangVersion>8</LangVersion>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.Bootstrap/BepInEx.Bootstrap.csproj:13]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Core/BepInEx.Core.csproj [hunk 2]
```diff
@@ -7,10 +7,6 @@
     <LangVersion>8</LangVersion>
     <Description>Unity plugin injection framework</Description>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.Core/BepInEx.Core.csproj:13]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Core/Bootstrap/BaseChainloader.cs [hunk 3]
```diff
@@ -343,10 +342,13 @@ protected static bool PluginTargetsWrongBepin(PluginInfo pluginInfo)
...
-			if (pluginTarget.Minor > CurrentAssemblyVersion.Minor) return true;
-			if (pluginTarget.Minor < CurrentAssemblyVersion.Minor) return false;
-			return pluginTarget.Build > CurrentAssemblyVersion.Build;
+			if (pluginTarget.Major != CurrentAssemblyVersion.Major)
+				return true;
+			if (pluginTarget.Minor > CurrentAssemblyVersion.Minor)
+				return true;
+			if (pluginTarget.Minor < CurrentAssemblyVersion.Minor)
+				return false;
+			return pluginTarget.Patch > CurrentAssemblyVersion.Build;
```
Comment (denikson) [inline BepInEx.Core/Bootstrap/BaseChainloader.cs:351]: Formatting this is unnecessary as it makes the diff less clear. We prefer doing formatting in separate commits/PRs so that the relevant diff is properly visible.

File: BepInEx.Core/Contract/Attributes.cs [hunk 4]
```diff
@@ -20,34 +23,35 @@ public class BepInPlugin : Attribute
...
+		/// <summary>
+		/// Loads metadata from assembly attributes
+		/// </summary>
+		/// <param name="guid">The unique identifier of the plugin. Should not change between plugin versions.</param>
+		public BepInPlugin(string guid)
 		{
-			this.GUID = GUID;
-			this.Name = Name;
+			this.GUID = guid;
+		}
```
Comment (denikson) [inline BepInEx.Core/Contract/Attributes.cs:43]: Why was this added and why does it need to be public?

File: BepInEx.Core/Contract/Attributes.cs [hunk 5]
```diff
@@ -57,7 +61,29 @@ internal static BepInPlugin FromCecilType(TypeDefinition td)
...
+				{
+					SemVersion.TryParse(versionAttribute.ConstructorArguments.Single().Value as string, out version);
+				}
+				else
+				{
+					version = new SemVersion(name.Version);
+				}
+
+				return new BepInPlugin(guid, name.Name, version);
+			}
```
Comment (denikson) [inline BepInEx.Core/Contract/Attributes.cs:84]: Was this meant to be used as follows:

```cs
[BepInPlugin("my.plugin.id")]
class MyPlugin : BasePlugin {}
```
and it would grab plugin version and name from assembly name and version?

To me sounds like an interesting idea, but I'm not sure if it's good to grab the name from assembly name. Name was initially intended as a human-readable normal name of the plugin and not really an assembly name.

cc @bbepis for opinions on this

File: BepInEx.IL2CPP/BepInEx.IL2CPP.csproj [hunk 6]
```diff
@@ -7,10 +7,6 @@
     <OutputPath>..\bin\IL2CPP\</OutputPath>
     <LangVersion>8</LangVersion>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.IL2CPP/BepInEx.IL2CPP.csproj:13]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.IL2CPP/BepInEx.IL2CPP.csproj [hunk 7]
```diff
@@ -53,11 +49,5 @@
     <PackageReference Include="Iced" Version="1.6.0" />
     <PackageReference Include="MonoMod.RuntimeDetour" Version="20.11.5.1" />
   </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Properties" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Properties\AssemblyInfo.cs" />
-  </ItemGroup>
```
Comment (denikson) [inline BepInEx.IL2CPP/BepInEx.IL2CPP.csproj:61]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.NetLauncher/BepInEx.NetLauncher.csproj [hunk 8]
```diff
@@ -7,10 +7,6 @@
     <Product>BepInEx.NetLauncher</Product>
     <LangVersion>8</LangVersion>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.NetLauncher/BepInEx.NetLauncher.csproj:13]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.NetLauncher/BepInEx.NetLauncher.csproj [hunk 9]
```diff
@@ -28,11 +24,5 @@
   <ItemGroup>
     <PackageReference Include="HarmonyX" Version="2.1.1" />
   </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Properties" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Properties\AssemblyInfo.cs" />
-  </ItemGroup>
```
Comment (denikson) [inline BepInEx.NetLauncher/BepInEx.NetLauncher.csproj:36]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Patcher/BepInEx.Patcher.csproj [hunk 10]
```diff
@@ -11,10 +11,6 @@
     <LangVersion>8</LangVersion>
     <CodeAnalysisRuleSet>MinimumRecommendedRules.ruleset</CodeAnalysisRuleSet>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.Patcher/BepInEx.Patcher.csproj:17]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Patcher/BepInEx.Patcher.csproj [hunk 11]
```diff
@@ -39,11 +35,5 @@
   <ItemGroup>
     <EmbeddedResource Include="..\bin\patcher\BepInEx.Bootstrap.dll" />
   </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Properties" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Properties\AssemblyInfo.cs" />
-  </ItemGroup>
```
Comment (denikson) [inline BepInEx.Patcher/BepInEx.Patcher.csproj:47]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Preloader.Core/BepInEx.Preloader.Core.csproj [hunk 12]
```diff
@@ -6,10 +6,6 @@
     <OutputPath>..\bin\</OutputPath>
     <LangVersion>8</LangVersion>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.Preloader.Core/BepInEx.Preloader.Core.csproj:12]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Preloader.Unity/BepInEx.Preloader.Unity.csproj [hunk 13]
```diff
@@ -6,10 +6,6 @@
     <OutputPath>..\bin\</OutputPath>
     <LangVersion>8</LangVersion>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.Preloader.Unity/BepInEx.Preloader.Unity.csproj:12]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Preloader.Unity/BepInEx.Preloader.Unity.csproj [hunk 14]
```diff
@@ -25,11 +21,5 @@
   <ItemGroup>
     <PackageReference Include="HarmonyX" Version="2.1.1" />
   </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Properties" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Properties\AssemblyInfo.cs" />
-  </ItemGroup>
```
Comment (denikson) [inline BepInEx.Preloader.Unity/BepInEx.Preloader.Unity.csproj:33]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Shared/BepInEx.Shared.projitems [hunk 15]
```diff
@@ -4,13 +4,19 @@
...
   <ItemGroup>
-    <Compile Include="$(MSBuildThisFileDirectory)AssemblyInfo.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)BuildInfoAttribute.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)VersionInfo.cs" />
+  </ItemGroup>
+  
+  <ItemGroup Condition="'$(BuildInfo)' != ''">
+    <AssemblyAttribute Include="BepInEx.Shared.BuildInfoAttribute">
+        <_Parameter1>$(BuildInfo)</_Parameter1>
+    </AssemblyAttribute>
```
Comment (denikson) [inline BepInEx.Shared/BepInEx.Shared.projitems:20]: Not related to semver, should be a separate PR.

File: BepInEx.Unity/BepInEx.Unity.csproj [hunk 16]
```diff
@@ -6,10 +6,6 @@
     <OutputPath>..\bin\Unity\</OutputPath>
     <LangVersion>8</LangVersion>
     <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
-    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
-    <GenerateAssemblyCopyrightAttribute>false</GenerateAssemblyCopyrightAttribute>
-    <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
-    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Comment (denikson) [inline BepInEx.Unity/BepInEx.Unity.csproj:12]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInEx.Unity/BepInEx.Unity.csproj [hunk 17]
```diff
@@ -32,11 +28,5 @@
   <ItemGroup>
     <PackageReference Include="HarmonyX" Version="2.1.1" />
   </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Properties" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Properties\AssemblyInfo.cs" />
-  </ItemGroup>
```
Comment (denikson) [inline BepInEx.Unity/BepInEx.Unity.csproj:40]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: BepInExTests/BepInExTests.csproj [hunk 18]
```diff
@@ -39,10 +39,4 @@
     <ProjectReference Include="..\BepInEx.IL2CPP\BepInEx.IL2CPP.csproj" />
     <ProjectReference Include="..\BepInEx.Unity\BepInEx.Unity.csproj" />
   </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Properties" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Properties\AssemblyInfo.cs" />
-  </ItemGroup>
```
Comment (denikson) [inline BepInExTests/BepInExTests.csproj:47]: How is this related to semver change? This should be in a separate PR for better testability and ability to revert it if needed.

File: build.cake [hunk 19]
```diff
@@ -1,8 +1,8 @@
-#addin nuget:?package=Cake.FileHelpers&version=3.2.1
```
Comment (denikson) [inline build.cake:1]: This whole cakebuild script change is nice, but it is not related to semver. Therefore it should be a separate PR.

**Review by denikson — CHANGES_REQUESTED** [review]
Thanks for the PR! Here are some notes. I already mentioned some of them on Discord, here so in some cases this is a reiteration of already mentioned things.

File: BepInEx.Core/Contract/Attributes.cs (hunk 5 above)
Comment (js6pak) [reply BepInEx.Core/Contract/Attributes.cs:84]: I could just use [AssemblyTitleAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assemblytitleattribute?view=netframework-3.5) with fallback to assembly name, just like AssemblyInformationalVersionAttribute with assembly version

**Review by denikson — APPROVED** [review]
Looks good to me

**Comment by denikson** [comment]
cc @bbepis for final verification on this PR

**Comment by denikson** [comment]
After some delibiration with Bepis, there was a new decision regarding this:

* Semver ranges are indeed needed: their immediate use is to update [`BepInDependency`](https://bepinex.github.io/bepinex_docs/master/api/BepInEx.BepInDependency.html#BepInEx_BepInDependency__ctor_System_String_System_String_) attribute to specify semver version range instead of a minimum version
    * Chainloading should be upgraded accordingly to do version range checks
* Currently used semver port to net35 doesn't have ranges while the original NuGet versioning package was too big (nor supported net35) either. Instead we should use [semver.net](https://github.com/adamreeve/semver.net) which supports version ranges and net35 out of the box
* There is no consensus yet as to how semver.net should be included, but for now it's clear that we'll pull it directly as a NuGet package

There may be a change related maybe ILRepacking semver into core, but for now it's fine just as a NuGet package since net35 version is available directly.

**Review by denikson — APPROVED** [review]
Looks good to me, waiting final approval of @bbepis

**Review by bbepis — APPROVED** [review]
